File: src/bumperbot_controller/src/twist_to_twist_stamped.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/twist_stamped.hpp"

class TwistToTwistStamped : public rclcpp::Node
{
public:
    TwistToTwistStamped()
        : Node("twist_to_twist_stamped")
    {
        // Subscription to /key_vel (Twist messages)
        subscription_ = this->create_subscription<geometry_msgs::msg::Twist>(
            "/key_vel", 10,
            std::bind(&TwistToTwistStamped::twistCallback, this, std::placeholders::_1));

        // Publisher to /input_joy/cmd_vel_stamped (TwistStamped messages)
        publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(
            "/input_joy/cmd_vel_stamped", 10);

        RCLCPP_INFO(this->get_logger(), "TwistToTwistStamped node has been started.");
    }

private:
    void twistCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
    {
        // Create a TwistStamped message
        auto twist_stamped = geometry_msgs::msg::TwistStamped();
        twist_stamped.header.stamp = this->get_clock()->now(); // Add the current timestamp
        twist_stamped.header.frame_id = "base_link";           // Optional: Set a frame ID
        twist_stamped.twist = *msg;                            // Copy the Twist message

        // Publish the TwistStamped message
        publisher_->publish(twist_stamped);
    }

    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr subscription_;
    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr publisher_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TwistToTwistStamped>());
    rclcpp::shutdown();
    return 0;
}

File: src/bumperbot_controller/bumperbot_controller/twist_to_twist_stamped.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist, TwistStamped
from std_msgs.msg import Header

class TwistToTwistStamped(Node):
    def __init__(self):
        super().__init__('twist_to_twist_stamped')
        self.subscription = self.create_subscription(
            Twist,
            '/key_vel',
            self.twist_callback,
            10)
        self.publisher = self.create_publisher(TwistStamped, '/input_joy/cmd_vel_stamped', 10)

    def twist_callback(self, msg):
        twist_stamped = TwistStamped()
        twist_stamped.header = Header()
        twist_stamped.header.stamp = self.get_clock().now().to_msg()
        twist_stamped.twist = msg
        self.publisher.publish(twist_stamped)

def main(args=None):
    rclpy.init(args=args)
    node = TwistToTwistStamped()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()